atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CqlClientImpl.java

```diff
@@ -0,0 +1,144 @@
+        private final BoundStatement boundStatement;
+
+        CqlQueryImpl(BoundStatement boundStatement, ResultSetVisitor<R> resultSetVisitor) {
+            this.boundStatement = boundStatement;
+            this.resultSetVisitor = resultSetVisitor;
+        }
+
+        final AsyncFunction<ResultSet, R> iterate() {
+            return rs -> {
+                Preconditions.checkArgument(rs != null, "ResultSet should not be null when iterating");
+
+                R currentResult = resultSetVisitor.visit(rs);
+
+                boolean wasLastPage = rs.getExecutionInfo().getPagingState() == null;
+                if (wasLastPage) {
+                    return Futures.immediateFuture(currentResult);
+                } else {
+                    ListenableFuture<ResultSet> future = rs.fetchMoreResults();
+                    return Futures.transformAsync(future, iterate(), executor);
+                }
+            };
+        }
+
+        @Override
+        public ListenableFuture<R> execute() {
+            return Futures.transformAsync(session.executeAsync(boundStatement), iterate(), executor);
+        }
+    }
+
+    @Override
+    public CqlQueryBuilder<Object> queryBuilder() {
+        return new CqlQueryBuilderImpl<>();
+    }
+
+    private class CqlQueryBuilderImpl<R> implements CqlQueryBuilder<R> {
+
+        private String queryString;
+        private Map<String, Object> args = new HashMap<>();
+        private ResultSetVisitor<R> resultSetVisitor;
+
+        @Override
+        public CqlQueryBuilder<R> setQueryString(String query) {
+            this.queryString = query;
+            return this;
+        }
+
+        @Override
+        public CqlQueryBuilder<R> setArg(String argumentName, Object argument) {
+            args.put(argumentName, argument);
+            return this;
+        }
+
+        @Override
+        public <T> CqlQueryBuilder<T> setResultSetVisitor(ResultSetVisitor<T> visitor) {
+            this.resultSetVisitor = (ResultSetVisitor<R>) visitor;
+            return (CqlQueryBuilder<T>) this;
+        }
+
+        @Override
+        public CqlQuery<R> build() {
+            Preconditions.checkState(queryString != null, "Empty query string");
+            PreparedStatement statement = CqlClientImpl.this.prepareStatement(queryString);
+
+            Preconditions.checkState(statement.getVariables().size() == args.size(),
+                    "Number of query variables and numbers of set variables are not the same");
+            Object[] argsArray = new Object[args.size()];
+            for (int i = 0; i < args.size(); i++) {
+                Preconditions.checkState(this.args.keySet().contains(statement.getVariables().getName(i)),
+                        "Set argument is not expected");
+                argsArray[i] = args.get(statement.getVariables().getName(i));
+            }
+
+            // a workaround to support different types for arguments
```

Comment:
I saw one comment which says `when we switch to Java 9 use foo()`, my goal with this comment was to have something similar so that if the driver starts supporting `setObject` we switch to that. Comment removed.